langlib/lang.function/src/main/java/org/ballerinalang/langlib/function/Call.java

```diff
@@ -0,0 +1,134 @@
+        List<Type> paramTypes = new LinkedList<>();
+        List<Type> argTypes = new LinkedList<>();
+        List<Object> argsList = new java.util.ArrayList<>(List.of(Scheduler.getStrand()));
+
+        if (validatePositionalArgs(args, argsList, functionType, paramTypes, argTypes) ||
+                validateRestArgs(args, argsList, functionType, paramTypes, argTypes)) {
+                throw ErrorCreator.createError(
+                        getModulePrefixedReason(FUNCTION_LANG_LIB, INCOMPATIBLE_TYPES),
+                        BLangExceptionHelper.getErrorDetails(RuntimeErrors.INCOMPATIBLE_TYPE,
+                                                             new BTupleType(paramTypes), new BTupleType(argTypes)));
+        }
+
+        return func.asyncCall(argsList.toArray(), METADATA);
+    }
+
+    private static boolean validatePositionalArgs(Object[] args, List<Object> argsList, BFunctionType functionType,
+                                                  List<Type> paramTypes, List<Type> argTypes) {
+        boolean errored = false;
+        Parameter[] parameters = functionType.parameters;
+        int numOfParams = parameters.length;
+        int numOfArgs = args.length;
+        for (int i = 0; i < numOfParams; i++) {
+            Parameter parameter = parameters[i];
+            Type paramType = parameter.type;
+            paramTypes.add(paramType);
+            if (i >= numOfArgs) {
+                if (parameter.isDefault) {
+                    argsList.add(0);
+                    argsList.add(false);
+                } else {
+                    errored = true;
+                }
+                continue;
+            }
+            Object arg = args[i];
+            Type argType = TypeChecker.getType(arg);
+            argTypes.add(argType);
+            if (!TypeChecker.checkIsType(null, arg, argType, paramType)) {
+                errored = true;
+            }
+            argsList.add(arg);
+            argsList.add(true);
+        }
+        return errored;
+    }
+
+    private static boolean validateRestArgs(Object[] args, List<Object> argsList, BFunctionType functionType,
+                                            List<Type> paramTypes, List<Type> argTypes) {
+        boolean errored = false;
+        int numOfArgs = args.length;
+        int numOfRestArgs = Math.max(numOfArgs - functionType.parameters.length, 0);
+        BArrayType restType = (BArrayType) functionType.restType;
+        if (restType != null) {
+            ListInitialValueEntry.ExpressionEntry[] initialValues =
+                                                    new ListInitialValueEntry.ExpressionEntry[numOfRestArgs];
+            Type elementType = restType.getElementType();
+            for (int i = 0; i < numOfRestArgs; i++) {
+                Object arg = args[numOfArgs - numOfRestArgs + i];
+                Type argType = TypeChecker.getType(arg);
+                paramTypes.add(elementType);
+                argTypes.add(argType);
```

Comment:
This will give a confusing error with something like

```ballerina
public function main() {
    any|error x = function:call(fn, 1, 2, 3);
}

function fn(int a, string... b) {

}
```

```cmd
error: {ballerina/lang.function}IncompatibleTypes {"message":"incompatible types: expected '[int,string,string]', found '[int,int,int]'"}
        at ballerina.lang.function.0:call(function.bal:27)
```

Kinda implies that there are 3 required params, right? Can we use the function type to build a param string representation that preserves the rest parameter?

Reply:
That's argument-specific. Here we are giving a param list vs arg list comparison. So it's not quite the same.